Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs

```diff
@@ -0,0 +1,271 @@
+using Content.Shared.Access.Components;
+using Content.Server.Radio.EntitySystems;
+using Content.Server.Power.Components;
+using Content.Shared.Chat;
+using Robust.Shared.Player;
+using Content.Server.PDA.Ringer;
+using Robust.Shared.Network;
+using Content.Server.DeviceNetwork;
+using Content.Server.DeviceNetwork.Components;
+using Content.Server.DeviceNetwork.Systems;
+using Content.Shared.DeviceNetwork;
+
+
+namespace Content.Server.CartridgeLoader.Cartridges;
+
+public sealed class MessagesCartridgeSystem : EntitySystem
+{
+    [Dependency] private readonly CartridgeLoaderSystem _cartridgeLoaderSystem = default!;
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly GameTicker _gameTicker = default!;
+    [Dependency] private readonly MessagesServerSystem _messagesServerSystem = default!;
+    [Dependency] private readonly DeviceNetworkSystem _deviceNetworkSystem = default!;
+
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<MessagesCartridgeComponent, CartridgeMessageEvent>(OnUiMessage);
+        SubscribeLocalEvent<MessagesCartridgeComponent, CartridgeUiReadyEvent>(OnUiReady);
+        SubscribeLocalEvent<MessagesCartridgeComponent, DeviceNetworkPacketEvent>(OnPacketReceived);
+    }
+
+    /// <summary>
+    /// This gets called when the ui fragment needs to be updated for the first time after activating
+    /// </summary>
+    private void OnUiReady(EntityUid uid, MessagesCartridgeComponent component, CartridgeUiReadyEvent args)
+    {
+        UpdateUiState(uid, args.Loader, component);
+    }
+
+    /// <summary>
+    /// The ui messages received here get wrapped by a CartridgeMessageEvent and are relayed from the <see cref="CartridgeLoaderSystem"/>
+    /// </summary>
+    /// <remarks>
+    /// The cartridge specific ui message event needs to inherit from the CartridgeMessageEvent
+    /// </remarks>
+    private void OnUiMessage(EntityUid uid, MessagesCartridgeComponent component, CartridgeMessageEvent args)
+    {
+        if (args is not MessagesUiMessageEvent messageEvent)
+            return;
+
+        if (messageEvent.Action == MessagesUiAction.Send && TryComp(uid, out CartridgeComponent? cartComponent) && GetUserUid(cartComponent) is int userId && component.ChatUid != null && messageEvent.StringInput != null)
+        {
+            MessagesMessageData messageData = new()
+            {
+                SenderId = userId,
+                ReceiverId = component.ChatUid.Value,
+                Content = messageEvent.StringInput,
+                Time = _gameTiming.CurTime.Subtract(_gameTicker.RoundStartTimeSpan)
+            };
+            if (!TryComp(uid, out DeviceNetworkComponent? device))
+                return;
+            var packet = new NetworkPayload()
+            {
+                ["Message"] = messageData
+            };
+            _deviceNetworkSystem.QueuePacket(uid, null, packet, device: device);
```

Comment:
Don't broadcast the packet. Use `_singletonServerSystem.TryGetActiveServerAddress` to get the address of the active server.